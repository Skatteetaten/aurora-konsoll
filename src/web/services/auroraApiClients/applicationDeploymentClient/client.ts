import GoboClient, { IDataAndErrors } from 'web/services/GoboClient';
import {
  REDEPLOY_WITH_CURRENT_VERSION_MUTATION,
  REDEPLOY_WITH_VERSION_MUTATION,
  REFRESH_APPLICATION_DEPLOYMENT_MUTATION,
  REFRESH_APPLICATION_DEPLOYMENTS_MUTATION,
  DELETE_APPLICATION_DEPLOYMENT_MUTATION,
  DeployResponse,
  UpdateAuroraConfigFileInput,
  AuroraConfigFileValidationResponse,
  UPDATE_AURORA_CONFIG_FILE,
  DeployCurrentResponse,
} from './mutation';
import {
  APPLICATION_DEPLOYMENT_DETAILS_QUERY,
  APPLICATIONS_QUERY,
  IApplicationDeploymentDetailsQuery,
  IApplicationsConnectionData,
  IUserAndAffiliationsData,
  USER_AFFILIATIONS_QUERY,
  IApplicationDeploymentWithDetailsData,
  APPLICATION_DEPLOYMENT_WITH_DETAILS_QUERY,
} from './query';

export class ApplicationDeploymentClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async updateAuroraConfigRedeployAndRefreshDeployment(
    updateAuroraConfigFileInput: UpdateAuroraConfigFileInput,
    applicationDeploymentId: string
  ) {
    const updateFileResult = await this.updateAuroraConfigFile(
      updateAuroraConfigFileInput
    );

    if (updateFileResult.data?.updateAuroraConfigFile.success) {
      const redeployResult = await this.redeployWithCurrentVersion(
        applicationDeploymentId
      );
      if (
        redeployResult.data?.redeployWithCurrentVersion?.applicationDeploymentId
      ) {
        await this.refreshApplicationDeployment(applicationDeploymentId);
        return await this.fetchApplicationDeploymentWithDetails(
          applicationDeploymentId
        );
      } else {
        return {
          name: redeployResult.name,
          errors: redeployResult.errors,
        };
      }
    } else {
      return {
        name: updateFileResult.name,
        errors: updateFileResult.errors,
      };
    }
  }

  public async redeployWithVersionAndRefreshDeployment(
    applicationDeploymentId: string,
    version: string
  ): Promise<IDataAndErrors<IApplicationDeploymentWithDetailsData>> {
    const redeployResult = await this.redeployWithVersion(
      applicationDeploymentId,
      version
    );
    if (redeployResult.data?.redeployWithVersion?.applicationDeploymentId) {
      await this.refreshApplicationDeployment(applicationDeploymentId);
      return await this.fetchApplicationDeploymentWithDetails(
        applicationDeploymentId
      );
    } else {
      return {
        name: redeployResult.name,
        errors: redeployResult.errors,
      };
    }
  }

  public async deleteAndRefreshApplications(
    affiliation: string,
    namespace: string,
    name: string
  ): Promise<IDataAndErrors<IApplicationsConnectionData>> {
    const deleteResponse = await this.deleteApplicationDeployment(
      namespace,
      name
    );
    if (
      deleteResponse.data &&
      deleteResponse.data.deleteApplicationDeployment
    ) {
      return await this.refreshAndFetchApplications([affiliation]);
    } else {
      return {
        name: deleteResponse.name,
        errors: deleteResponse.errors,
      };
    }
  }

  public async refreshAndFetchApplicationDeployment(
    applicationDeploymentId: string
  ): Promise<IDataAndErrors<IApplicationDeploymentWithDetailsData>> {
    const refreshResult = await this.refreshApplicationDeployment(
      applicationDeploymentId
    );
    if (refreshResult.data && refreshResult.data.refreshApplicationDeployment) {
      return await this.fetchApplicationDeploymentWithDetails(
        applicationDeploymentId
      );
    } else {
      return {
        name: refreshResult.name,
        errors: refreshResult.errors,
      };
    }
  }

  public async fetchApplicationDeploymentWithDetails(
    applicationDeploymentId: string
  ): Promise<IDataAndErrors<IApplicationDeploymentWithDetailsData>> {
    return await this.client.query<IApplicationDeploymentWithDetailsData>({
      query: APPLICATION_DEPLOYMENT_WITH_DETAILS_QUERY,
      variables: {
        id: applicationDeploymentId,
      },
    });
  }

  public async redeployWithVersion(
    applicationDeploymentId: string,
    version: string
  ): Promise<IDataAndErrors<DeployResponse>> {
    return await this.client.mutate<DeployResponse>({
      mutation: REDEPLOY_WITH_VERSION_MUTATION,
      variables: {
        input: {
          applicationDeploymentId,
          version,
        },
      },
    });
  }

  public async redeployWithCurrentVersion(
    applicationDeploymentId: string
  ): Promise<IDataAndErrors<DeployCurrentResponse>> {
    return await this.client.mutate<DeployCurrentResponse>({
      mutation: REDEPLOY_WITH_CURRENT_VERSION_MUTATION,
      variables: {
        input: {
          applicationDeploymentId,
        },
      },
    });
  }

  public async refreshApplicationDeployment(
    applicationDeploymentId: string
  ): Promise<IDataAndErrors<{ refreshApplicationDeployment: boolean }>> {
    return await this.client.mutate<{
      refreshApplicationDeployment: boolean;
    }>({
      mutation: REFRESH_APPLICATION_DEPLOYMENT_MUTATION,
      variables: {
        input: {
          applicationDeploymentId,
        },
      },
    });
  }

  public async refreshAndFetchApplications(
    affiliations: string[]
  ): Promise<IDataAndErrors<IApplicationsConnectionData>> {
    const refreshResponse = await this.refreshAffiliations(affiliations);
    if (
      refreshResponse.data &&
      refreshResponse.data.refreshApplicationDeployments
    ) {
      return await this.findAllApplicationDeployments(affiliations);
    } else {
      return {
        name: refreshResponse.name,
        errors: refreshResponse.errors,
      };
    }
  }

  public async refreshAffiliations(
    affiliations: string[]
  ): Promise<IDataAndErrors<{ refreshApplicationDeployments: boolean }>> {
    return await this.client.mutate<{
      refreshApplicationDeployments: boolean;
    }>({
      mutation: REFRESH_APPLICATION_DEPLOYMENTS_MUTATION,
      variables: {
        input: {
          affiliations,
        },
      },
    });
  }

  public async findApplicationDeploymentDetails(
    id: string
  ): Promise<IDataAndErrors<IApplicationDeploymentDetailsQuery> | undefined> {
    return await this.client.query<IApplicationDeploymentDetailsQuery>({
      query: APPLICATION_DEPLOYMENT_DETAILS_QUERY,
      variables: {
        id,
      },
    });
  }

  public async findUserAndAffiliations(): Promise<
    IDataAndErrors<IUserAndAffiliationsData>
  > {
    return await this.client.query<IUserAndAffiliationsData>({
      query: USER_AFFILIATIONS_QUERY,
    });
  }

  public async findAllApplicationDeployments(
    affiliations: string[]
  ): Promise<IDataAndErrors<IApplicationsConnectionData>> {
    return await this.client.query<IApplicationsConnectionData>({
      query: APPLICATIONS_QUERY,
      variables: {
        affiliations,
      },
    });
  }

  public async deleteApplicationDeployment(
    namespace: string,
    name: string
  ): Promise<IDataAndErrors<{ deleteApplicationDeployment: boolean }>> {
    return await this.client.mutate<{ deleteApplicationDeployment: boolean }>({
      mutation: DELETE_APPLICATION_DEPLOYMENT_MUTATION,
      variables: {
        input: {
          namespace,
          name,
        },
      },
    });
  }

  public async updateAuroraConfigFile({
    auroraConfigName,
    auroraConfigReference,
    fileName,
    contents,
    existingHash,
  }: UpdateAuroraConfigFileInput) {
    return await this.client.mutate<AuroraConfigFileValidationResponse>({
      mutation: UPDATE_AURORA_CONFIG_FILE,
      variables: {
        input: {
          auroraConfigName,
          auroraConfigReference,
          fileName,
          contents,
          existingHash,
        },
      },
    });
  }
}
