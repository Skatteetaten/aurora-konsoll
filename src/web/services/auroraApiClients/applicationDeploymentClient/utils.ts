import YAML from 'yaml';

export function changeVersionInFile(
  fileName: string,
  fileContent: string,
  version: string
) {
  if (fileName.endsWith('json')) {
    const parsedApplicationFile: JSON = JSON.parse(fileContent);
    return JSON.stringify(
      { ...parsedApplicationFile, version: version },
      null,
      2
    );
  }
  try {
    const parsedApplicationFile = YAML.parseDocument(fileContent);
    parsedApplicationFile.set('version', version);
    return parsedApplicationFile.toString();
  } catch {
    return undefined;
  }
}
