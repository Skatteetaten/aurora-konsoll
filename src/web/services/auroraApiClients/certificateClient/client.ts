import { ICertificatesQuery } from 'web/models/certificates';
import GoboClient, { IDataAndErrors } from 'web/services/GoboClient';
import { CERTIFICATES_QUERY } from './query';

export class CertificateClient {
  public client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }
  public async getCertificates(): Promise<
    IDataAndErrors<ICertificatesQuery> | undefined
  > {
    return await this.client.query<ICertificatesQuery>({
      query: CERTIFICATES_QUERY,
    });
  }
}
