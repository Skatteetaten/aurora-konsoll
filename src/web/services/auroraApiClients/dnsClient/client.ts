import GoboClient, { IDataAndErrors } from 'web/services/GoboClient';
import { CnameQuery, CNAME_QUERY } from './query';

export class DnsClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async fetchCname(
    affiliation: string
  ): Promise<IDataAndErrors<CnameQuery>> {
    return await this.client.query<CnameQuery>({
      query: CNAME_QUERY,
      variables: {
        affiliation,
      },
    });
  }
}
