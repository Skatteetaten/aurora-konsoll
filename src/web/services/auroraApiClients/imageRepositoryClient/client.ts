import GoboClient, { IDataAndErrors } from 'web/services/GoboClient';
import { ITagsQuery, VERSIONS_QUERY, ITagQuery, TAG_QUERY } from './query';

interface IImageTagsVariables {
  repositories: string[];
  types?: string[];
}

export class ImageRepositoryClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async fetchTag(
    repository: string,
    tagName: string
  ): Promise<IDataAndErrors<ITagQuery>> {
    return await this.client.query<ITagQuery>({
      query: TAG_QUERY,
      variables: {
        repositories: [repository],
        names: [tagName],
      },
    });
  }

  public async findVersions(
    repository: string
  ): Promise<IDataAndErrors<ITagsQuery>> {
    let variables: IImageTagsVariables = {
      repositories: [repository],
    };

    return await this.client.query<ITagsQuery>({
      query: VERSIONS_QUERY,
      variables,
    });
  }

  public async fetchVersions(
    repository: string
  ): Promise<IDataAndErrors<ITagsQuery>> {
    return this.findVersions(repository);
  }
}
