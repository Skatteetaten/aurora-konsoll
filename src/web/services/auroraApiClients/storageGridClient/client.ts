import GoboClient, { IDataAndErrors } from 'web/services/GoboClient';
import {
  ObjectAreas,
  StorageGridObjectArea,
  StorageGridQuery,
  STORAGEGRID_AREAS_QUERY,
  STORAGEGRID_TENANT_QUERY,
  Tenant,
} from './query';

export interface AreasAndTenant {
  activeAreas?: StorageGridObjectArea[];
  isTenantRegistered?: boolean;
}

export class StorageGridClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getAreasAndTenant(
    affiliation: string
  ): Promise<IDataAndErrors<AreasAndTenant>> {
    const tenant = await this.getTenant(affiliation);

    const isTenantRegistered =
      tenant.data?.affiliations.edges[0].node.storageGrid.tenant.isRegistered;

    if (isTenantRegistered === true) {
      const areas = await this.getAreas(affiliation);
      const activeAreas =
        areas.data?.affiliations.edges[0].node.storageGrid.objectAreas.active;

      return {
        name: areas.name,
        data: {
          isTenantRegistered,
          activeAreas: activeAreas,
        },
        errors: areas.errors,
      };
    }

    return {
      name: tenant.name,
      data: {
        isTenantRegistered,
        activeAreas: undefined,
      },
      errors: tenant.errors,
    };
  }

  public async getAreas(
    affiliation: string
  ): Promise<IDataAndErrors<StorageGridQuery<ObjectAreas>>> {
    return await this.client.query<StorageGridQuery<ObjectAreas>>({
      query: STORAGEGRID_AREAS_QUERY,
      variables: { affiliation },
    });
  }

  public async getTenant(
    affiliation: string
  ): Promise<IDataAndErrors<StorageGridQuery<Tenant>>> {
    return await this.client.query<StorageGridQuery<Tenant>>({
      query: STORAGEGRID_TENANT_QUERY,
      variables: { affiliation },
    });
  }
}
