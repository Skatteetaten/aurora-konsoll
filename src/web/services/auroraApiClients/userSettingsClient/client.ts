import { IUserSettings } from 'web/models/UserSettings';
import GoboClient, { IDataAndErrors } from 'web/services/GoboClient';
import { UPDATE_USERSETTINGS_MUTATION } from './mutation';
import { IUserSettingsData, USERSETTINGS_QUERY } from './query';

export class UserSettingsClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getUserSettings(): Promise<IDataAndErrors<IUserSettingsData>> {
    return await this.client.query<IUserSettingsData>({
      query: USERSETTINGS_QUERY,
    });
  }

  public async updateUserSettings(
    userSettings: IUserSettings
  ): Promise<IDataAndErrors<IUserSettingsData>> {
    const result = await this.client.mutate<{
      updateUserSettings: boolean;
    }>({
      mutation: UPDATE_USERSETTINGS_MUTATION,
      variables: {
        input: userSettings,
      },
    });

    if (result.data && result.data.updateUserSettings) {
      return await this.getUserSettings();
    } else {
      return {
        name: result.name,
        errors: result.errors,
      };
    }
  }
}
