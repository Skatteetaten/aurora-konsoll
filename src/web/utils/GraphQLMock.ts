import cors from 'cors';
import express from 'express';
import { Server } from 'http';
import GoboClient from 'web/services/GoboClient';
import { IApiClients } from 'web/models/AuroraApi';
import {
  ApplicationDeploymentClient,
  CertificateClient,
  DatabaseClient,
  ImageRepositoryClient,
  NetdebugClient,
  UserSettingsClient,
  WebsealClient,
} from 'web/services/auroraApiClients';
import { Socket } from 'net';
import { DnsClient } from 'web/services/auroraApiClients/dnsClient/client';

type ResponseMock = any;

interface IGraphQLResponseMock {
  [queryName: string]: ResponseMock;
}

export class GraphQLSeverMock {
  public port: number;
  public graphQLUrl: string;
  private server: Server;
  private responses: IGraphQLResponseMock[];
  private serverSockets = new Set<Socket>();

  constructor(port?: number) {
    const PORT = port || randomPort(40000, 50000);
    const ENDPOINT = '/graphql';

    const server = express();
    server.use(cors());
    server.use(express.json());

    server.post(ENDPOINT, (req, res) => {
      const { operationName } = req.body;
      const responseMock = this.responses[operationName];
      if (!responseMock) {
        res.sendStatus(404);
      } else {
        res.send(responseMock);
      }
    });

    this.responses = [];
    this.server = server.listen(PORT);
    this.port = PORT;
    this.graphQLUrl = `http://localhost:${PORT}${ENDPOINT}`;

    // Keeps track of open connections that must be closed after test is done.
    this.server.on('connection', (socket) => {
      this.serverSockets.add(socket);
      socket.on('close', () => {
        this.serverSockets.delete(socket);
      });
    });
  }

  /**
   * Closes all connections and the server.
   */
  public close(done: jest.DoneCallback) {
    this.server.close(() => {
      done();
    });
    for (const socket of this.serverSockets.values()) {
      socket.destroy();
    }
  }

  public putResponse(queryName: string, response: any) {
    this.responses[queryName] = response;
  }
}

function randomPort(start: number, stop: number): number {
  const random = Math.random() * (stop - start) + start;
  return Math.trunc(random);
}

export function goboClientMock(url: string): GoboClient {
  return new GoboClient({ url });
}

export function getApiClientsMock(serverMock: GraphQLSeverMock): IApiClients {
  const clientMock = goboClientMock(serverMock.graphQLUrl);
  return {
    applicationDeploymentClient: new ApplicationDeploymentClient(clientMock),
    certificateClient: new CertificateClient(clientMock),
    databaseClient: new DatabaseClient(clientMock),
    imageRepositoryClient: new ImageRepositoryClient(clientMock),
    netdebugClient: new NetdebugClient(clientMock),
    userSettingsClient: new UserSettingsClient(clientMock),
    websealClient: new WebsealClient(clientMock),
    dnsClient: new DnsClient(clientMock),
  };
}
