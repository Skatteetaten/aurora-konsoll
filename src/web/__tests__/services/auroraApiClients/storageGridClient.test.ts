import { StorageGridClient } from 'web/services/auroraApiClients/storageGridClient/client';
import { goboClientMock, GraphQLSeverMock } from 'web/utils/GraphQLMock';

import * as getTenant from './__responses__/storageGridClient/getTenant.json';
import * as getAreas from './__responses__/storageGridClient/getAreas.json';

const serverMock = new GraphQLSeverMock();
const clientMock = goboClientMock(serverMock.graphQLUrl);
const storageGridClient = new StorageGridClient(clientMock);

afterAll((done) => {
  serverMock.close(done);
});

describe('getTenant', () => {
  it('should fetch tenant', async () => {
    serverMock.putResponse('getTenant', getTenant);

    const result = await storageGridClient.getTenant('paas');
    expect(result).toMatchSnapshot();
  });
});

describe('getAreas', () => {
  it('should fetch areas', async () => {
    serverMock.putResponse('getAreas', getAreas);

    const result = await storageGridClient.getAreas('paas');
    expect(result).toMatchSnapshot();
  });
});
