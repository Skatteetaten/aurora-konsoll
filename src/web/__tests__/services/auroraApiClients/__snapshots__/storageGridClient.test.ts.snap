// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAreas should fetch areas 1`] = `
Object {
  "data": Object {
    "affiliations": Object {
      "edges": Array [
        Object {
          "node": Object {
            "storageGrid": Object {
              "objectAreas": Object {
                "active": Array [
                  Object {
                    "bucketName": "bucket",
                    "creationTimestamp": "2020-01-22T17:30:42Z",
                    "name": "test",
                    "namespace": "paas-test",
                    "status": Object {
                      "message": "Failed",
                      "reason": "Unknown",
                      "success": false,
                    },
                  },
                ],
              },
            },
          },
        },
      ],
    },
  },
  "errors": undefined,
  "name": "getAreas",
}
`;

exports[`getTenant should fetch tenant 1`] = `
Object {
  "data": Object {
    "affiliations": Object {
      "edges": Array [
        Object {
          "node": Object {
            "storageGrid": Object {
              "tenant": Object {
                "isRegistered": true,
              },
            },
          },
        },
      ],
    },
  },
  "errors": undefined,
  "name": "getTenant",
}
`;
